nemo/collections/nlp/models/language_modeling/megatron_ptune_gpt_model.py

```diff
@@ -0,0 +1,523 @@
+        self.embeddings = self.model.model.language_model.embedding.word_embeddings
+
+        # self.vocab = self.tokenizer.tokenizer.get_vocab()
+
+        self.template = cfg.prompt_encoder.template
+
+        self.prompt_encoder = PromptEncoder(
+            template=cfg.prompt_encoder.template,
+            hidden_size=hidden_size,
+            lstm_dropout=cfg.prompt_encoder.dropout,
+            num_layers=cfg.prompt_encoder.num_layers,
+        )
+
+        # load prompt encoder
+        self.hidden_size = hidden_size
+        self.tokenizer.add_special_tokens({'additional_special_tokens': [cfg.pseudo_token]})
+
+        self.pseudo_token_id = self.tokenizer.token_to_id(cfg.pseudo_token)
+        self.pad_token_id = self.tokenizer.pad_id if self.tokenizer.pad_id is not None else self.tokenizer.unk_id
+        self.spell_length = sum(self.template)
+        self.special_tokens = set(
+            [
+                self.tokenizer.eos_id,
+                self.tokenizer.pad_id,
+                self.tokenizer.sep_id,
+                self.tokenizer.unk_id,
+                self.tokenizer.bos_id,
+                self.tokenizer.cls_id,
+            ]
+        )
+
+    def embed_input(self, queries, enc_taskname):
+        bz = queries.shape[0]
+        queries_for_embedding = queries.clone()
+
+        queries_for_embedding[(queries == self.pseudo_token_id)] = self.pad_token_id
+
+        raw_embeds = self.embeddings(queries_for_embedding).clone()
+        if self.cfg.prompt_encoder.task_dependent:
+            enc_taskname = self.embeddings(enc_taskname)
+        else:
+            enc_taskname = None
+
+        if self.float_type == torch.float32:
+            replace_embeds = self.prompt_encoder(enc_taskname=enc_taskname)
+        else:
+            with torch.autocast(device_type="cuda", dtype=self.float_type):
+                replace_embeds = self.prompt_encoder(enc_taskname=enc_taskname)
+
+        blocked_indices = queries == self.pseudo_token_id
+        raw_embeds = raw_embeds.clone().type(self.float_type)
+        # find the index to the psedo-tokens
+        index = blocked_indices.nonzero().reshape((bz, -1, 2))[:, :, 1][:, :, None]
+
+        _, seq, _ = index.shape
+        _, _, emb = raw_embeds.shape
+        index = index.expand(bz, seq, emb)
+
+        if enc_taskname is None:
+            # taskname none, encoder returens batch 1
+            # need to expand
+            _, replace_seq, _ = replace_embeds.shape
+            replace_embeds = replace_embeds.expand(bz, replace_seq, emb)
+
+        # scatter the psedo-token embeddings to the raw embeddings
+        raw_embeds.scatter_(1, index, replace_embeds)
+        # slow version of above scatter logics
+        # for bidx in range(bz):
+        #     position = blocked_indices[bidx].nonzero()[:, 0]
+        #     for i in range(len(position)):
+        #         raw_embeds[bidx, position[i], :] = replace_embeds[bidx, i, :]
```

Comment:
clean up comments

Reply:
I will leave it since it helps people to understand the code.